tool/yubico-piv-tool.c

```diff
@@ -2218,14 +2219,14 @@ int main(int argc, char *argv[]) {
   OPENSSL_init_crypto(OPENSSL_INIT_LOAD_CONFIG, 0);
 #endif
 
-  if(ykpiv_init(&state, verbosity) != YKPIV_OK) {
-    fprintf(stderr, "Failed initializing library.\n");
+  if((rc = ykpiv_init(&state, verbosity)) != YKPIV_OK) {
+    fprintf(stderr, "Failed initializing library: %s.\n", ykpiv_strerror(rc));
     cmdline_parser_free(&args_info);
     return EXIT_FAILURE;
   }
 
-  if(ykpiv_connect(state, args_info.reader_arg) != YKPIV_OK) {
-    fprintf(stderr, "Failed to connect to yubikey.\nTry removing and reconnecting the device.\n");
+  if((rc = ykpiv_connect(state, args_info.reader_arg)) != YKPIV_OK) {
+    fprintf(stderr, "Failed to connect to yubikey: %s.\nTry removing and reconnecting the device.\n", ykpiv_strerror(rc));
```

Comment:
If the service is down it will say this:

Failed to connect to yubikey: PCSC service not available.
Try removing and reconnecting the device.

So it is clear that the service is down. Do you find this too unclear ?

Reply:
Yes I see the second part of the message is a little misleading in that case, but perhaps the better solution would be to remove that text entirely. A non-technical user would not know how to go about fixing things if the PCSC service is not running.

Reply:
And to your comment: The first part is the error message, which correctly identifies the service as unavailable. The second part is just some generic text that tries to be helpful. I don't think printing re-start the service is certain to be true in the first place, and won't helt an average user much either. How do you restart the PCSC service ...

Reply:
To be helpful message, if the error is `YKPIV_PCSC_SERVICE_ERROR` the generic helpful message would be something like `try to restart the pcsc service`, otherwise the current helpful message. :)

Reply:
Yeah, one solution would be to simply make that text part of the message defined in error.c

Reply:
I've fixed the stuff we discussed above, as well as improved the error handling in the pkcs11 module (well error debug logging I should say), and added a second commit to this PR.